### PR TITLE
[dtensor][10/n] switch pointwise op to use op strategy

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111278
* __->__ #111234
* #110717
* #109145

As titled, this also handles sth like [Shard(0), Shard(0)] correctly for
pointwise ops, which was previously errored out

### CODE REVIEW

File: torch/distributed/_tensor/ops/pointwise_ops.py [hunk 1]
```diff
@@ -376,11 +401,116 @@
 ]
 
 
+def pointwise_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    print(f">>>>>>>> op input op schema: {op_schema}")
```
Comment (fduwjj) [inline torch/distributed/_tensor/ops/pointwise_ops.py:405]: Is this needed?

File: torch/distributed/_tensor/ops/pointwise_ops.py [hunk 2]
```diff
@@ -376,11 +401,116 @@
...
+            PlacementStrategy(
+                output_spec=DTensorSpec(
+                    mesh=mesh,
+                    placements=tuple(out_placements),
+                ),
+                input_specs=input_specs,
+                redistribute_cost=redistribute_costs,
+            )
+        )
+    print(f">>>>>>> final pointwise strategy generated: {pointwise_strategy}.")
```
Comment (fduwjj) [inline torch/distributed/_tensor/ops/pointwise_ops.py:502]: ditto.

File: torch/distributed/_tensor/ops/pointwise_ops.py [hunk 3]
```diff
@@ -376,11 +401,116 @@
...
 for op in linear_pointwise_ops:
     register_prop_rule(op)(linear_pointwise_rule)
 
 
 for op in pointwise_ops:
-    register_prop_rule(op, schema_info=RuntimeSchemaInfo(static_kwargkey=["out"]))(
-        pointwise_rule
+    # register_prop_rule(op, schema_info=RuntimeSchemaInfo(static_kwargkey=["out"]))(
+    #     pointwise_rule
+    # )
```
Comment (fduwjj) [inline torch/distributed/_tensor/ops/pointwise_ops.py:513]: maybe we want to remove them?

File: torch/distributed/_tensor/ops/pointwise_ops.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/_tensor/ops/pointwise_ops.py]: updated

File: torch/distributed/_tensor/ops/pointwise_ops.py [hunk 4]
```diff
@@ -376,11 +394,109 @@
...
+                )
+                input_specs.append(
+                    DTensorSpec(
+                        mesh=mesh,
+                        placements=input_placements,
+                        tensor_meta=input_arg_spec.tensor_meta,
+                    )
+                )
+                redistribute_costs.append(
+                    generate_redistribute_costs(input_arg, spec_to_follow)
```
Comment (fduwjj) [inline torch/distributed/_tensor/ops/pointwise_ops.py:479]: Wait, IIUC, you want to use `input_placements` to calculate `redistribute_cost`?

File: torch/distributed/_tensor/ops/pointwise_ops.py [hunk 5]
```diff
@@ -376,11 +394,109 @@
...
+        follow_operand_dims_map = infer_broadcast_dims_map(
+            common_shape, followed_strategy.output_shape
+        )
+
+    assert isinstance(
+        followed_strategy, OpStrategy
+    ), f"no strategy to follow for {op_schema}!"
+    pointwise_strategy = OpStrategy([])
+
+    for placement_strategy in followed_strategy.strategies:
```
Comment (fduwjj) [inline torch/distributed/_tensor/ops/pointwise_ops.py:438]: Not a strong opinion here, but for better code readability, is it possible that we give the `followed_strategy` a better like, IIUC, this kind of implies how output (op result) can be distributed? For my learning purpose, why do we call it "follow"?

Comment (fduwjj) [reply torch/distributed/_tensor/ops/pointwise_ops.py:438]: And also, under what circumstances when there is multiple strategies for a followed_strategy?

**Review by fduwjj — APPROVED** [review]
Unblock for now, mainly some questions.

File: torch/distributed/_tensor/ops/pointwise_ops.py (hunk 4 above)
Comment (wanchaol) [reply torch/distributed/_tensor/ops/pointwise_ops.py:479]: Yeah good catch

File: torch/distributed/_tensor/ops/pointwise_ops.py (hunk 5 above)
Comment (wanchaol) [reply torch/distributed/_tensor/ops/pointwise_ops.py:438]: "follow" means all other args should follow this strategy. There might be cases there're multiple strategies for a graph mode case, for eager there's no case that there's multiple strategies

**Pull request closed without merging**